### DISCUSSION TITLE
What is the proper way to change h1 color (for example)

### DISCUSSION THREAD

**Discussion by ohabash** [tailwindlabs/tailwindcss#4460] [Help]
### HTML

```
<div class="copy-block prose-xl prose-red prose">
                <h1>POTENTIAL MADE POSSIBLE</h1>
                <p class="f">For 30 years, companies have relied on Habitat Commons to empower them to achieve their full potential by providing strategic and creative solutions for growth. We are focused on creating results through partnership, expertise and innovation.For 30 years, companies have relied on Habitat Commons to empower them to achieve their full potential by providing
                strategic and creative solutions for growth. We are focused on creating results through partnership, expertise and
                innovation.</p>
            </div>
`````

The color assigned to that h1 is `text-gray-900` (I dont know why).

### CSS
I am trying to find the best way to set up paragraphs headers etc. So this is my attempt (not working)

```
h1 {
    @apply text-cyan;
}
```

that does not show up in the dom anywhere. 

### My stack
- angular 12
- scss
- nx workspace (nrwl)
- tailwindcss: 2.1.2

----------

### **I have two questions** that could clear this up.
*I am new to TW and I am trying to do everything the "new" way but it is very tempting just to go back to old ways.* 
1. What is the recommended way to define typography sizes // line-height // font-family etc?
2. where is `text-gray-900` defined as header text color? (do you recommend anything to read?)

**Comment by johanchouquet** [comment]
Hi @ohabash, this is because you have the `prose` class in the parent `div`, which sets for you quite a few classes for a bunch of child elements. If you code in VSCode, you'll be able to see which ones.
You can have theses features just by adding the `prose` class, which is enabled if you have load the `tailwind/typography` plugin in your config.

For the rest, I'll leave others answer you, as I'm interested in the answers myself.

**Reply by therealgilles** [reply]
@johanchouquet, do you mean `prose` instead of `prone`?

**Reply by johanchouquet** [reply]
Thanks for noticing that @therealgilles . Updated my answer.

**Comment by simonswiss** [comment]
Hello @ohabash 👋

Welcome to the community! 🎉 I'll try my best to answer your questions here.

## What is the recommended way to define typography sizes // line-height // font-family etc?

The [Adding base styles page of the Tailwind CSS documentation website](https://tailwindcss.com/docs/adding-base-styles) is a great  overview of the different options available to achieve that:

### Global typographic styles

If you want to apply default typography styles to your entire project (target raw HTML elements like `h1`, `blockquote` etc in your CSS), the recommendation is to inject styles [into your `@base` layer](https://tailwindcss.com/docs/adding-base-styles#using-css).

In the CSS file where you defined the Tailwind directives...

```
@tailwind base;
@tailwind components;
@tailwind utilities;
```

...you can add an `@layer base {}` block, which will place the CSS you write inside of it in the correct source order inside the generated CSS output:

```
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer base {
  h1 {
    @apply text-gray-800;
  }
}
```

There are definitely good use cases for doing so, but before you jump ahead and define base styles for everything like that, consider adopting a different mindset and apply utility classes directly on the elements in your markup. 

If you'd rather define global typographic styles, that's fine and the `@base` layer would be one way to go about it. Just make sure you take some time to consider the paradigm shift that Tailwind proposes. It might seem silly, at first, to repeat yourself with utility classes in your markup. 

Try and challenge that gut reaction - it's a good onboarding exercise. Perhaps consider making the abstraction at the markup level (reusable component, template partial), which is another way to stay DRY 👍

####  Side note about the `text-cyan` class you're trying to apply 

In your example, you try to use `h1 { @apply text-cyan; }`. This won't work (with default colors anyway), because there is no color called just `cyan`.  The `cyan` palette gives you 10 shades: `cyan-50`, `cyan-100`, `cyan-200`, ..., `cyan-800`, `cyan-900`.

If you wanted to `@apply` a cyan color, you'd need to pick one of those shades:

```css
h1 {
  @apply text-cyan-800
}
```

Also, `cyan` is not part of the [default color palette](https://tailwindcss.com/docs/customizing-colors), so it will need to be imported and extended in your Tailwind config file:

```diff

// tailwind.config.js
const colors = require('tailwindcss/colors')

module.exports = {
  theme: {
+    extend: {
+      colors: {
+        cyan: colors.cyan
+      }
+    }
  }
}
```



### Scoped typographic styles with the `@tailwindcss/typography` plugin

In some cases, you don't want to apply typographic styles globally, but only have them _scoped_ to a specific portion of your project. That is really useful in scenarios where you'd have some content coming from an external source (WYSIWYG field in CMS, API response, etc) where you cannot control the markup and apply utility classes to it directly.

There is an official Tailwind CSS plugin for that – the [Typography plugin](https://github.com/tailwindlabs/tailwindcss-typography).

Based on the code example you have shared, this is what your project is using. That plugins gives you a `prose` parent class, which will scope typographic styles to anything within the element it is applied to.

## Where is text-gray-900 defined as header text color? (do you recommend anything to read?)

> The color assigned to that h1 is text-gray-900 (I dont know why).

In your example, your `h1` has the `text-gray-900` color because it sits inside an element with the `prose` class. And [the default styles for the `h1` element](https://github.com/tailwindlabs/tailwindcss-typography/blob/master/src/styles.js#L93-L96) inside a `prose` wrapper is set to be `text-gray-900`.

If you're going to work with this Typography plugin, I suggest you check this short video (by me 😎) dedicated on this plugin:

https://www.youtube.com/watch?v=J0Wy359NJPM

It will show you one of the possible ways to modify/extend/override the default styles generated by the plugin.

## I hope it helps!

Pfeeewwww, that was a long-winded response! I hope this is useful, and please let me know if it makes sense, and if you have additional questions.

> I am trying to do everything the "new" way but it is very tempting just to go back to old ways.

I hear you. I'd bet every single person using Tailwind has been there too. It's an _almost required first step_ in your journey into this controversial paradigm shift.

My pro-tip: stick with it, give it a real try for a week or two, and see what happens. Chances are you'll view things from a completely different angle then.

And if you still don't like it, that's perfectly fine too. At least you'll know why! 🤗

Good luck, and let me know how you go! ✨

**Comment by ohabash** [comment]
ah this is good info. Thank you! I watched the video and that cleared things up. I had actually already watched several of ur videos over the past week. you're famous.

I think a way to globally define client typography once is important for a few reasons. There will be several other devs, paying way less attention to design than is necessary to carry out things like

```
<h1 class="text-green font-bold text-5xl mb-4 leading-3">Page Title</h1>
```

I think it invites everyone to make changes to core design. extending base style seems like the better choice.

Simple `<h1>Page Title</h1>`  is necessary.

```
h1 {
    @apply text-green font-bold text-5xl mb-4 leading-3;
}
```

But i agree this does not follow paradigm.  And the theme is not as portable anymore. That is the main reason we chose Tailwind. 

Is there a way to use your typography plugin to achieve site wide styles? I put `prose` on `<body>` but its not working currently as expected.



being able to extend theme configs like this would be amazing.

```
      typography: {
        extend: {
          h1: [
            "text-green",
            "font-bold",
            "text-5xl",
            "mb-4",
            "leading-3"],
          label: [
            'uppercase',
            "opacity-75",
            "tracking-wider"
          ],
          ".main-box": [
            "py-8",
            "text-base",
            "leading-6",
            "space-y-4",
          ],
          ul: [
            "list-disc",
            "space-y-2",
            { li: [ "flex", "items-start" ] }
          ]
        }
      }
```

in a micro frontend app or in a customizable saas product this would be gold; because client themes [_**and typography**_] would be created with frontend theme generators and stored in client databases so each client will have there own look and feel across many frontends.

**Comment by tnorthcutt** [comment]
I'd like to add something to what @simonswiss said:

> There are definitely good use cases for doing so, but before you jump ahead and define base styles for everything like that, consider adopting a different mindset and apply utility classes directly on the elements in your markup.
> 
> If you'd rather define global typographic styles, that's fine and the @base layer would be one way to go about it. Just make sure you take some time to consider the paradigm shift that Tailwind proposes. It might seem silly, at first, to repeat yourself with utility classes in your markup.
> 
> Try and challenge that gut reaction - it's a good onboarding exercise. Perhaps consider making the abstraction at the markup level (reusable component, template partial), which is another way to stay DRY 👍

Another way to think about this is that Tailwind works really well with components. I'm not familiar with Angular, but I'm guessing it uses components much like React or Vue. If you're building with components, then you won't be writing the same `<h1 class="text-red-500">Heading here</h1>` over and over; instead you'd probably have something like `<h1 class="text-red-500>{{ title }}</h1>` and pass `title` in as a prop to the component (or however Angular does this).

Granted, there will of course be cases where you're writing new markup, but I wanted to share that perspective. If you find yourself writing the same elements with the same classes over and over, that's probably a good candidate for extracting to a component, regardless of whether you're using Tailwind or not 😃

**Reply by ohabash** [reply]
This is a good point. I have been making reusable components for the purpose of keeping markup simple.

However... A new component for `h1` / `h2` / `p` is not reasonable . So far — _good-old-fashioned_ css is the winner here. (example below)... I really think we need the ability to add base styles through the tailwind config. 

```
body {
    @apply text-fg;
}

h1, h2 {
    @apply font-bold text-5xl mb-4 uppercase;
    line-height: 1.2em;
}

h2 {
    @apply text-6xl font-extrabold;
}

p {
    &.f {
        @apply font-semibold text-base opacity-75 text-lg;
        line-height: 1.85em;
    }
}
```

Would love to hear if there is a better way or if i am missing something. (I am very new to tailwind)

**Reply by tnorthcutt** [reply]
Yep I agree, if you're making components for individual HTML elements you've gone too far 😀

I think what you're doing seems reasonable, and most importantly you will probably start to notice if there are opportunities to improve your setup as you get more comfortable with Tailwind.

Enjoy! I love using Tailwind and consider it the best way to handle CSS.
